lib/sage.ex

```diff
@@ -437,6 +447,57 @@ defmodule Sage do
   def run_async(sage, name, transaction, compensation, opts \\ []),
     do: add_stage(sage, name, build_operation!(:run_async, transaction, compensation, opts))
 
+  @doc """
+  For a given Sage S with transactions :t1 -> :t2 -> :t3, a call to `interleave(S, :name, f)`
+  will yield a saga with transactions :t1 -> :name_1 -> :t2 -> :name_2 -> :t3 -> :name_3.
+
+  This can be useful if you are trying to do a long computation and want to do something with
+  the intermediate results, such as logging or persistence.
+  
+  Note: 
+  - This isn't strict interleaving because a transaction is still appended at the end.
+  - Calling this function twice with the same name will give a `Sage.DuplicateStageError`, but
+  calling the function twice with a different name will compound its effects.
+  - Calling this function before the end of your saga definition will mean any stages you add after the
+  call to `interleave` will not have the intermediate stages added after them
+  """
+  @spec interleave(
+          sage :: t(),
+          name :: stage_name(),
+          intermediate_transaction :: intermediate_transaction(),
+          compensation :: compensation()
+        ) :: t()
+  def interleave(sage, name, intermediate_transaction, compensation \\ :noop)
+      when is_intermediate_transaction(intermediate_transaction) and is_compensation(compensation) do
+    sage.stages
+    |> Enum.reverse()
+    |> Enum.with_index(1)
+    |> Enum.flat_map(fn {stage, index} ->
+      name = String.to_atom("#{name}_#{index}")
+      {stage_name, _} = stage
+
+      transaction =
+        case intermediate_transaction do
+          {m, f, a} ->
+            {m, f, [stage_name | a]}
+
+          _ ->
+            &intermediate_transaction.(&1, &2, stage_name)
+        end
+
+      [stage, {name, build_operation!(:run, transaction, compensation)}]
+    end)
+    |> Enum.reduce(new(), fn stage, sage ->
```

Comment:
I think you don't need to reconstruct the Sage completely, just update the existing one like so: `%{sage | steps: steps}`. Struct can be extended in the future and we don't want to re-apply everything else that is in there or lose it if we forget about it.

Reply:
I took a stab at a refactor, but I'm not sold on it. I am leaning towards thinking that the best approach is to rebuild the saga and refactor later if need be, but let me know what you think of the refactor attempt.

Reply:
@apreifsteck if we use the suggestion below and namespace the interleaves then there can't be a collisition with the regular stage names, so we can do something like this:
`%{sage | stage_names: MapSet.add(sage.stage_names, {:interleave, name}), steps: steps)`
then just add a manual check for another interleave not colliding with the `{:interleave, name}` right in this function.

Reply:
I've implemented your suggestion to use tuples as the stage names. I'm not sure if I understand the advantage of adding a check in this function to look for duplicate stage names. That seems like a separate responsibility that `interleave` and `run` could both take advantage of.